File: API/2890_Blau_TVI_Timed_Reversal/CS/BlauTviTimedReversalStrategy.cs
using System;
using System.Collections.Generic;

using Ecng.Common;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// Blau TVI timed reversal strategy (simplified). Uses Momentum indicator
/// to detect slope changes and generate reversal entries.
/// </summary>
public class BlauTviTimedReversalStrategy : Strategy
{
	private readonly StrategyParam<DataType> _candleType;
	private readonly StrategyParam<int> _momentumLength;

	public DataType CandleType
	{
		get => _candleType.Value;
		set => _candleType.Value = value;
	}

	public int MomentumLength
	{
		get => _momentumLength.Value;
		set => _momentumLength.Value = value;
	}

	public BlauTviTimedReversalStrategy()
	{
		_candleType = Param(nameof(CandleType), TimeSpan.FromHours(4).TimeFrame())
			.SetDisplay("Candle Type", "Candles", "General");

		_momentumLength = Param(nameof(MomentumLength), 12)
			.SetGreaterThanZero()
			.SetDisplay("Momentum Length", "Momentum period", "Indicators");
	}

	protected override void OnStarted2(DateTime time)
	{
		base.OnStarted2(time);

		var momentum = new Momentum { Length = MomentumLength };

		decimal prevMom = 0;
		decimal prevPrevMom = 0;
		var count = 0;

		var subscription = SubscribeCandles(CandleType);
		subscription
			.Bind(momentum, (ICandleMessage candle, decimal momValue) =>
			{
				if (candle.State != CandleStates.Finished)
					return;

				count++;
				if (count < 3)
				{
					prevPrevMom = prevMom;
					prevMom = momValue;
					return;
				}

				if (!IsFormedAndOnlineAndAllowTrading())
				{
					prevPrevMom = prevMom;
					prevMom = momValue;
					return;
				}

				// Detect slope reversal: was falling, now rising => buy
				if (prevMom < prevPrevMom && momValue > prevMom && Position <= 0)
					BuyMarket();
				// Was rising, now falling => sell
				else if (prevMom > prevPrevMom && momValue < prevMom && Position >= 0)
					SellMarket();

				prevPrevMom = prevMom;
				prevMom = momValue;
			})
			.Start();

		var area = CreateChartArea();
		if (area != null)
		{
			DrawCandles(area, subscription);
			DrawIndicator(area, momentum);
			DrawOwnTrades(area);
		}
	}
}

File: API/2890_Blau_TVI_Timed_Reversal/PY/blau_tvi_timed_reversal_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import Momentum
from StockSharp.Algo.Strategies import Strategy


class blau_tvi_timed_reversal_strategy(Strategy):
    def __init__(self):
        super(blau_tvi_timed_reversal_strategy, self).__init__()

        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromHours(4))) \
            .SetDisplay("Candle Type", "Candles", "General")
        self._momentum_length = self.Param("MomentumLength", 12) \
            .SetDisplay("Momentum Length", "Momentum period", "Indicators")

        self._prev_mom = 0.0
        self._prev_prev_mom = 0.0
        self._count = 0

    @property
    def CandleType(self):
        return self._candle_type.Value

    @property
    def MomentumLength(self):
        return self._momentum_length.Value

    def OnReseted(self):
        super(blau_tvi_timed_reversal_strategy, self).OnReseted()
        self._prev_mom = 0.0
        self._prev_prev_mom = 0.0
        self._count = 0

    def OnStarted(self, time):
        super(blau_tvi_timed_reversal_strategy, self).OnStarted(time)
        self._prev_mom = 0.0
        self._prev_prev_mom = 0.0
        self._count = 0

        momentum = Momentum()
        momentum.Length = self.MomentumLength

        subscription = self.SubscribeCandles(self.CandleType)
        subscription.Bind(momentum, self._on_process).Start()

        area = self.CreateChartArea()
        if area is not None:
            self.DrawCandles(area, subscription)
            self.DrawIndicator(area, momentum)
            self.DrawOwnTrades(area)

    def _on_process(self, candle, mom_value):
        if candle.State != CandleStates.Finished:
            return
        mv = float(mom_value)
        self._count += 1
        if self._count < 3:
            self._prev_prev_mom = self._prev_mom
            self._prev_mom = mv
            return
        if not self.IsFormedAndOnlineAndAllowTrading():
            self._prev_prev_mom = self._prev_mom
            self._prev_mom = mv
            return
        if self._prev_mom < self._prev_prev_mom and mv > self._prev_mom and self.Position <= 0:
            self.BuyMarket()
        elif self._prev_mom > self._prev_prev_mom and mv < self._prev_mom and self.Position >= 0:
            self.SellMarket()
        self._prev_prev_mom = self._prev_mom
        self._prev_mom = mv

    def CreateClone(self):
        return blau_tvi_timed_reversal_strategy()
